linkerd/distribute/src/cache.rs

```diff
@@ -0,0 +1,73 @@
+use super::{params, NewDistribute};
+use linkerd_stack::{layer, NewService, Param};
+use parking_lot::Mutex;
+use std::{fmt::Debug, hash::Hash, sync::Arc};
+
+/// A [`NewService`] that produces [`NewDistribute`]s using a shared cache of
+/// backends.
+#[derive(Debug)]
+pub struct CacheNewDistribute<K, N, S>(Arc<Inner<K, N, S>>);
+
+#[derive(Debug)]
+pub struct Inner<K, N, S> {
+    new_backend: N,
+    backends: Mutex<ahash::AHashMap<K, S>>,
```

Comment:
AFAICT, the intention is to use this as the `NewService` passed to a `SpawnWatch` middleware, in which case, this mutex is never actually contended?

Reply:
yyyes though i think there's actually somewhat of a bug right now, since this cache is currently not scoped to the logical target